test/pipelinerun_test.go

```diff
@@ -45,6 +45,146 @@ var (
+				t.Fatalf("Failed to create Task `%s`: %s", getName(taskName, index), err)
+			}
+
+			if _, err := c.PipelineClient.Create(getHelloWorldPipelineWithSingularTask(index, namespace)); err != nil {
+				t.Fatalf("Failed to create Pipeline `%s`: %s", getName(pipelineName, index), err)
+			}
+		},
+		expectedTaskRuns:       []string{task1Name},
+		expectedNumberOfEvents: 0,
+		pipelineRunFunc:        getHelloWorldPipelineRunWithTimeout,
+	}}
+
+	for i, td := range tds {
+		t.Run(td.name, func(t *testing.T) {
+			td := td
+			t.Parallel()
+			// Note that getting a new logger has the side effect of setting the global metrics logger as well,
+			// this means that metrics emitted from these tests will have the wrong test name attached. We should
+			// revisit this if we ever start using those metrics (maybe use a different metrics gatherer).
+			logger := getContextLogger(t.Name())
+			logger.Info("Starting test ", td.name)
+			c, namespace := setup(t, logger)
+
+			knativetest.CleanupOnInterrupt(func() { tearDown(t, logger, c, namespace) }, logger)
+			defer tearDown(t, logger, c, namespace)
+
+			logger.Infof("Setting up test resources for %q test in namespace %s", td.name, namespace)
+			td.testSetup(t, c, namespace, i)
+
+			prName := fmt.Sprintf("%s%d", pipelineRunName, i)
+			_, err := c.PipelineRunClient.Create(td.pipelineRunFunc(i, namespace))
+			if err != nil {
+				t.Fatalf("Failed to create PipelineRun `%s`: %s", prName, err)
+			}
+
+			logger.Infof("Waiting for PipelineRun %s in namespace %s to complete", prName, namespace)
+			if err := WaitForPipelineRunState(c, prName, pipelineRunTimeout, PipelineRunFinished(), "PipelineRunSuccess"); err != nil {
+				t.Fatalf("Error waiting for PipelineRun %s to finish: %s", prName, err)
+			}
+
+			logger.Infof("Making sure the expected TaskRuns %s were created", td.expectedTaskRuns)
+			actualTaskrunList, err := c.TaskRunClient.List(metav1.ListOptions{LabelSelector: fmt.Sprintf("tekton.dev/pipelineRun=%s", prName)})
+			if err != nil {
+				t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", prName, err)
+			}
+			expectedTaskRunNames := []string{}
+			for _, runName := range td.expectedTaskRuns {
+				taskRunName := strings.Join([]string{prName, runName}, "-")
+				// check the actual task name starting with prName+runName with a random suffix
+				for _, actualTaskRunItem := range actualTaskrunList.Items {
+					if strings.HasPrefix(actualTaskRunItem.Name, taskRunName) {
+						taskRunName = actualTaskRunItem.Name
+					}
+				}
+				expectedTaskRunNames = append(expectedTaskRunNames, taskRunName)
+				r, err := c.TaskRunClient.Get(taskRunName, metav1.GetOptions{})
+				if err != nil {
+					t.Fatalf("Couldn't get expected TaskRun %s: %s", taskRunName, err)
+				}
+
+				for {
+					if r.Status.GetCondition(duckv1alpha1.ConditionSucceeded).IsUnknown() {
+						time.Sleep(100 * time.Millisecond)
+					} else {
+						break
+					}
+				}
```

Comment:
hmmmm i dont think we should need to do this sleep here, on line 123 we are waiting for the pipelinerun to complete, so if the taskruns are not completed at this point, that's a bug :thinking:

Reply:
changed to be a *WaitForTaskRunState.*, is not a bug because the pipeline/task is retried so a second wait condition is needed.